src/azure-cli-core/azure/cli/core/commands/arm.py

```diff
@@ -100,7 +100,7 @@ def build_parameters(self):
 
 def handle_template_based_exception(ex):
     try:
-        raise CLIError(ex.inner_exception.error.message)
+        raise CLIError(ex.inner_exception.error)
     except AttributeError:
```

Comment:
This is weird. Shouldn't `AttributeError` be captured by this line? I couldn't repro it. I simply deployed to a non-existing resource group: 
```
$ az group deployment create -g foo --template-file template.json 
Azure Error: ResourceGroupNotFound
Message: Resource group 'jltemprg1' could not be found.
```

Reply:
No, i think because the exception is within a raise() call.
Perhaps
```
try:
  msg = ex.inner_exception.error.message
  raise(msg)
except:
```
might have caught it as intended?

Reply:
If `AttributeError` is caught, it shouldn't trigger the call stack you mentioned...

Reply:
Good point. Not sure of the specifics of exception handling in python (particularly in constructors) but perhaps
```
  msg = ex.inner_exception.error.message
  raise CLIError(msg)
```
would behave differently.

If so, that is kind of disturbing. I will try it out time permitting.

Reply:
Or try something like
```
    raise CLIError("test".message)
```

Reply:
Or can you just run the command I used on a non-existing resource group and see if you can hit the exception? `az group deployment create -g foo --template-file template.json`

Reply:
Nope. I do not hit the exception. There is something different with that error.

Reply:
The real issue is inside raise CLIError() then:
```
Traceback (most recent call last):
  File "/opt/az/lib/python3.6/site-packages/knack/cli.py", line 206, in invoke
    cmd_result = self.invocation.execute(args)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 608, in execute
    raise ex
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 666, in _run_jobs_serially
    results.append(self._run_job(expanded_arg, cmd_copy))
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/__init__.py", line 657, in _run_job
    cmd_copy.exception_handler(ex)
  File "/opt/az/lib/python3.6/site-packages/azure/cli/core/commands/arm.py", line 110, in handle_template_based_exception
    raise CLIError(ex)
knack.util.CLIError: <unprintable CLIError object>
```

Reply:
```Traceback (most recent call last):
  File "/opt/az/lib/python3.6/logging/__init__.py", line 992, in emit
    msg = self.format(record)
  File "/opt/az/lib/python3.6/site-packages/knack/log.py", line 78, in format
    msg = logging.StreamHandler.format(self, record)
  File "/opt/az/lib/python3.6/logging/__init__.py", line 838, in format
    return fmt.format(record)
  File "/opt/az/lib/python3.6/logging/__init__.py", line 575, in format
    record.message = record.getMessage()
  File "/opt/az/lib/python3.6/logging/__init__.py", line 336, in getMessage
    msg = str(self.msg)
  File "/opt/az/lib/python3.6/site-packages/msrestazure/azure_exceptions.py", line 193, in __str__
    return str(self.error)
  File "/opt/az/lib/python3.6/site-packages/msrestazure/azure_exceptions.py", line 120, in __str__
    return error_bytes.decode('ascii')
UnicodeDecodeError: 'ascii' codec can't decode byte 0xe2 in position 907: ordinal not in range(128)
```

Reply:
```
def handle_template_based_exception(ex):
    try:
        print ("inner_exception: ", ex.inner_exception.__dict__)
        print ("_message:", ex.inner_exception._message)
        print ("message:", ex.inner_exception.error.message)
        raise CLIError(ex.inner_exception.error.message)
    except AttributeError:
        print ("raising exception: ", ex.__dict__)
        sys.exit(1)
        raise CLIError(ex)
```
results in
Case 1: over quota:
```
nner_exception:  {'error': 'InvalidTemplateDeployment', '_message': "The template deployment 'vm_deploy_Pb7nNKeXowfPauMqHNRrEgwljwfGCWpi' is not valid according to the validation procedure. The tracking id is '0113e58a-53d1-4873-aa68-058c63905f44'. See inner errors for details.", 'request_id': None, 'error_time': None, 'target': None, 'details': [<msrestazure.azure_exceptions.CloudErrorData object at 0x7f521e351438>], 'innererror': None, 'additionalInfo': None, 'data': None}
_message: The template deployment 'vm_deploy_Pb7nNKeXowfPauMqHNRrEgwljwfGCWpi' is not valid according to the validation procedure. The tracking id is '0113e58a-53d1-4873-aa68-058c63905f44'. See inner errors for details.
raising exception:  {'deserializer': <msrest.serialization.Deserializer object at 0x7f521e629fd0>, 'error': <msrestazure.azure_exceptions.CloudErrorData object at 0x7f521e351978>, 'message': "The template deployment 'vm_deploy_Pb7nNKeXowfPauMqHNRrEgwljwfGCWpi' is not valid according to the validation procedure. The tracking id is '0113e58a-53d1-4873-aa68-058c63905f44'. See inner errors for details.", 'response': <Response [400]>, 'status_code': 400, 'request_id': '0113e58a-53d1-4873-aa68-058c63905f44', 'inner_exception': <msrestazure.azure_exceptions.CloudErrorData object at 0x7f521e351978>}
```
Case 2: missing group:
```
inner_exception:  {'error': 'ResourceGroupNotFound', '_message': "Resource group 'foo' could not be found.", 'request_id': None, 'error_time': None, 'target': None, 'details': None, 'innererror': None, 'additionalInfo': None, 'data': None}
_message: Resource group 'foo' could not be found.
raising exception:  {'deserializer': <msrest.serialization.Deserializer object at 0x7f3bd8acfeb8>, 'error': <msrestazure.azure_exceptions.CloudErrorData object at 0x7f3bd8a95390>, 'message': "Resource group 'foo' could not be found.", 'response': <Response [404]>, 'status_code': 404, 'request_id': 'a504de6d-a15c-4c6e-bf8f-217b9dc6496e', 'inner_exception': <msrestazure.azure_exceptions.CloudErrorData object at 0x7f3bd8a95390>}
```

Reply:
Indeed, when dumping error_bytes in `azure_exceptions.py` using
```
        error_bytes = error_str.encode()
        try:
            return error_bytes.decode('ascii')
        except:
            return str(error_bytes)
```
I get:
```
...
Support/Parameters/%7B%22subId%22:%2210a34459-7c41-4f7d-bf8f-4fb31696d960%22,%22pesId%22:%2206bfd9d3-516b-d5c6-5802-169c800dec89%22,%22supportTopicId%22:%22e12e3d1d-7fa0-af33-c6d0-3c50df9658a3%22%7D by specifying parameters listed in the \xe2\x80\x98Details\xe2\x80\x99 section for deployment to succeed. Please read more about quota limits at https://docs.microsoft.com/en-us/azure/azure-supportability/per-vm-quota-requests.
```
Note the `\xe2`

Reply:
Changing the line to
```
error_bytes.decode('utf8')
```
fixes the problem for real.

Reply:
The smart single quotes around the word 'Details' are causing the problem. They're not ASCII.

Reply:
The proper fix is to eschew non-ascii chars in error messages. But since that is from upstream, that won't happen. 

As long as people (MS, Apple, lol) insist on using non-ascii the only solution is https://github.com/Azure/azure-cli/issues/12199#issuecomment-588350758